### PR TITLE
adapter/persist: Make columns of Materialized Views nullable

### PR DESCRIPTION
**Pull request by ParkMyCar**
This PR should allow us to use Persist's `compare_and_evolve_schema`, it does a few things:

1. Makes all columns of a Materialized View nullable, except those annotated with an `ASSERT NOT NULL`. This fixes a bug we saw previously where across an upgrade of Materialize columns could change their nullability.
2. Renamed the existing `schemas` field in Persist State to `deprecated_schemas` and added a new `schemas` field. This should cause us to ignore all of the existing schemas which have incorrect nullability and register new ones.
3. Updated the Catalog upgrade-check to now also check that the newly parsed items have the same schemas that we have in Persist. This should allow us to catch issues like https://github.com/MaterializeInc/incidents-and-escalations/issues/117 before the new version rolls out.
4. Added a `proptest` to make sure these new nullable schemas can still read any structured non-nullable data that has already been written.
5. Re-added the 'persist_schema_register' and 'persist_schema_require' dyncfgs since we're essentially re-starting the process.

Notably what this PR doesn't do but we talked about previously is also making all of the columns in a Table nullable, because AFAICT that is already the [existing behavior](https://github.com/MaterializeInc/materialize/blob/e47f5d2f5a1328ad6577e83c6a387550bca377ab/src/sql/src/plan/statement/ddl.rs#L268-L290). In planning we mark all of the columns nullable, unless they're annotated with `NOT NULL` or `PRIMARY KEY`, which seems like exactly our desired end state.

**Rollout**: Because we're essentially adding a new field I believe we'll need to turn ~~both~~ `persist_schema_register` ~~and `persist_schema_require`~~ off. When 0.119.0 finishes rolling out we can turn `persist_schema_register` on, and then in 0.120.0 we're guaranteed to have all schemas registered for all shards again, ~~at which point we can turn `persist_schema_require` back on~~. I might be misunderstanding the flow of how schemas get registered though, so @danhhz I would appreciate your thoughts here.
Update: Chatted with Dan an we shouldn't need to touch `persist_schema_require`.

### Motivation

Fixes https://github.com/MaterializeInc/incidents-and-escalations/issues/117

### Tips for reviewers

Everything is separated by commit, so you should be able to review the changes independently. 

For commit 2 where I added the 'deprecated_schemas' field, I would appreciate the most thorough review here. I essentially just added the new field where `rustc` told me to, and I think it's all correct but I'm not totally positive.

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered. ([trigger-ci for additional test/nightly runs](https://trigger-ci.dev.materialize.com/))
- [x] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [x] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [x] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [x] If this PR includes major [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note), I have pinged the relevant PM to schedule a changelog post.

### CODE REVIEW

File: src/persist-client/src/internal/state.proto [hunk 1]
```diff
@@ -202,11 +202,15 @@ message ProtoRollup {
...
 
     ProtoInlinedDiffs diffs = 17;
 
     // MIGRATION: We previously stored rollups as a `SeqNo -> string Key` map,
     // but now the value is a `struct HollowRollup`.
     map<uint64, string> deprecated_rollups = 12;
+    // MIGRATION: We previously registered schemas that have since changed
+    // nullability. We've made changes to stabilize the schemas and are just
+    // forgetting all of the old values.
+    reserved 18;
```
Comment (danhhz) [inline src/persist-client/src/internal/state.proto:215]: Hmm, I think we have to do this also for StateDiff. And then once that's done, I worry about getting "state diff does not apply cleanly errors" if we just forget the old fields. It's counter-intuitive I know, but I do think this is less risky if we rename all the old fields, but keep all the diff apply logic for them

Comment (danhhz) [reply src/persist-client/src/internal/state.proto:215]: Yeah, probably safest to add them both back. Looks like they still exist in LD. I'd probably just pull the old definitions out of git history and then not even bother rolling them out slowly, just default to on

**Comment by bkirwi** [comment]
Heads up that this is different than what I'd expected based on our conversations last week - in my memory we'd decided that tables etc. (and nested records) would also have columns marked nullable, since marking columns non-nullable provides no value but prevents us from ever making that column nullable in the future.

(Totally possible I misunderstood or you just have an intermediate state here, but figured I'd better mention just in case.)

**Comment by ParkMyCar** [comment]
> Heads up that this is different than what I'd expected based on our conversations last week - in my memory we'd decided that tables etc. (and nested records) would also have columns marked nullable, since marking columns non-nullable provides no value but prevents us from ever making that column nullable in the future.
> 
> (Totally possible I misunderstood or you just have an intermediate state here, but figured I'd better mention just in case.)

It's a great callout! Just updated the description with my thinking here, but basically `plan_create_table` already seems to have our ideal behavior, everything is nullable except columns marked `NOT NULL` or `PRIMARY KEY`.

**Comment by ParkMyCar** [comment]
> Ah right 🤦! Yeah, we have to default to `persist_schema_register` off for the initial rollout because of persist forward compatibility. `persist_schema_require` is inside the check for register, so I think that one can default to on?

You're totally right! The bit I was scared about [is here](https://github.com/MaterializeInc/materialize/pull/29779/commits/7a674568ceacea66666f5c63862814d16f38c1d1#diff-d944f8663d340226ae581007acff2fe3a754902d7052eebb3bf3f0f3a4fa6183R730-R733) where we used to assert a schema existed, but since we return a default we should be fine to keep `persist_schema_require` turned to `on`.

> Ideally, we'd have instructions for the release eng to flip the flag and restart all the envds during the maintenance window, so we don't have to wait a week to verify that everything works okay

Do you think it needs to be a full environment restart, or is the restart of `environmentd` that happens are part of 0dt enough? Wondering if we can thread the needle here a bit. If not I'm okay with just turning on `persist_schema_register` after the rollout has completed and just going through the normal flow

**Comment by ggevay** [comment]
Note that this will regress the performance of some customer dataflows. Probably not by much, but it's hard to be certain. I'll check at least the slt plan changes after they are in the diff to see whether there are any big regressions there. If no big regressions there, then this should give us some confidence that there also won't be big regressions at customers.

**Review by ggevay — CHANGES_REQUESTED** [review]
I've run the slt rewrite locally ([commit, feel free to cherry-pick to this PR](https://github.com/ggevay/materialize/commit/2702925f6eb9174b7c4e6c5406d16b949d8e03f3)), and unfortunately there are some blocker plan regressions:
- (There are many minor regressions where a new `Filter ... IS NOT NULL` appears. This is just a minor CPU regression, so it's kind of ok.)
- (There are some minor regressions for variadic outer joins, where a union with a collection that just has a constant null row appears. This is ok.)
- There is a significant regression in ldbc query 04 (`ldbc_bi.slt`), where somehow a new join input appears, with a new `ArrangeBy` and a new `Distinct`.
- There is a catastrophic regression in ldbc query 07, where a new cross join appears. (I classify plan regressions as catastrophic where it would cause a hard to resolve incident, e.g. a user would need to scale up multiple replica sizes, or maybe even no amount of scaling up would help. When a new cross join appears, then often no amount of scaling up helps, because cross joins are not scalable, because they are 100% skewed to one CPU core.)

I'll investigate why exactly these regressions are happening, and try to think whether there is some optimizer tweak to prevent them, but this might be hard. But before that, could you please explain it a bit why this change is necessary? Why is it a problem for Persist if the nullability of an MV column changes across a version upgrade?

If we really need to do this change, then one thing we could do is run an EXPLAIN on all customer plans, and check for significant or catastrophic plan regressions. If we are lucky, maybe there would be only acceptable regressions at customers. But this would be a tedious procedure, so it would be good to just not do this change if we could avoid it somehow.

**Review by jkosh44 — COMMENTED** [review]
Adapter parts LGTM, I'll wait for others to officially sign off

**Comment by ParkMyCar** [comment]
Converted to draft because we're going to take a different approach here

**Pull request closed without merging**